File: app/src/main/java/com/vladiknt/calcnt/Calculate.java
package com.vladiknt.calcnt;

import java.util.ArrayList;

public class Calculate {

    public static String result(String input) {
        String result;
        try {
            result = calcFunc(new String[]{"0", input})[1];
        } catch (Exception math) {
            return "error";
        }

        if ((long)(Double.parseDouble(result) * 100000000) == (long)Double.parseDouble(result) * 100000000)
            result = String.valueOf((int)Double.parseDouble(result));
        else if (((long)(Double.parseDouble(result) * 100000000))/100000000.0 != Double.parseDouble(result))
            result = String.valueOf(((long)(Double.parseDouble(result) * 100000000))/100000000.0);

        if (result.equals(String.valueOf(9.223372036854776E10)))
            result = "very much";

        return result;
    }

    private static String[] /* first - i, second - result*/ calcFunc(String[] input /*first - i, second - expression*/) {
        ArrayList<String> list = new ArrayList<String>();
        String number = ""; // if we have a part of number, we`ll add it there and get a full number after all
        int i = Integer.parseInt(input[0]);
        String str = input[1];
        while (i < str.length()) {
            if (str.charAt(i) == ')') {
                i++;
                break;
            }
            if (str.charAt(i) != '(') {
                switch (str.charAt(i)) {
                    case '+':
                    case '-':
                    case '×':
                    case '÷':
                    case '^':
                        list.add("" + str.charAt(i));
                        break;
                    default:
                        try {
                            if (str.charAt(i + 1) != '.')
                                Integer.parseInt("" + str.charAt(i + 1));
                            // if next char is digit
                            number += str.charAt(i);
                        } catch (Exception e) {
                            // if next char isn`t digit
                            number += str.charAt(i);
                            list.add(number);
                            number = "";
                        }
                        break;
                }
                i++;
            } else {
                String[] output = calcFunc(new String[]{String.valueOf(i + 1), str});
                i = Integer.parseInt(output[0]);
                list.add(output[1]);
            }
        }

        // calculating from list
        int iterator = 0;
        double a, b;
        // ^ firstly
        while (true) {
            if (iterator == list.size()) {
                iterator = 0;
                break;
            }
            if (list.get(iterator).equals("^")) {
                b = Double.parseDouble(list.remove(iterator + 1));
                list.remove(iterator);
                a = Double.parseDouble(list.remove(iterator - 1));
                list.add(iterator - 1, String.valueOf(Math.pow(a, b)));
                iterator = 0;
            }
            iterator++;
        }
        // × and ÷
        while (true) {
            if (iterator == list.size()) {
                iterator = 0;
                break;
            }
            if (list.get(iterator).equals("×")) {
                b = Double.parseDouble(list.remove(iterator + 1));
                list.remove(iterator);
                a = Double.parseDouble(list.remove(iterator - 1));
                list.add(iterator - 1, String.valueOf(a * b));
                iterator = 0;
            }
            if (list.get(iterator).equals("÷")) {
                b = Double.parseDouble(list.remove(iterator + 1));
                list.remove(iterator);
                a = Double.parseDouble(list.remove(iterator - 1));
                list.add(iterator - 1, String.valueOf(a / b));
                iterator = 0;
            }
            iterator++;
        }
        // + and -
        while (true) {
            if (list.size() == 1)
                break;
            if (list.get(iterator).equals("+")) {
                b = Double.parseDouble(list.remove(iterator + 1));
                list.remove(iterator);
                a = Double.parseDouble(list.remove(iterator - 1));
                list.add(iterator - 1, String.valueOf(a + b));
                iterator = 0;
            }
            if (list.get(iterator).equals("-")) {
                b = Double.parseDouble(list.remove(iterator + 1));
                list.remove(iterator);
                a = Double.parseDouble(list.remove(iterator - 1));
                list.add(iterator - 1, String.valueOf(a - b));
                iterator = 0;
            }
            iterator++;
        }
        return new String[]{String.valueOf(i), list.remove(0)};
    }

}

File: app/src/main/java/com/vladiknt/calcnt/Calculate.kt
package com.vladiknt.calcnt

import java.lang.Exception
import kotlin.math.pow

object Calculate {

    fun result(input: String): String {
        var result = ""
        try {
            result = core(Pair("0", input)).second

            if (result == "NaN")
                result = "error"
            // Rounding to int if result is something like 4.0
            else if ((result.toDouble() * 100000000).toLong() == result.toDouble().toLong() * 100000000)
                result = (result.toDouble().toInt()).toString()
            // To avoid situations like "0.1 + 0.2 = 0.3000000000004"
            else if ((result.toDouble() * 100000000).toLong() / 100000000.0 != result.toDouble())
                result = ((result.toDouble() * 100000000).toLong() / 100000000.0).toString()

            if (result == "9.223372036854776E10")
                result = "very much"
        } catch (math: Exception) {
            return "error"
        }

        return result
    }

    private fun core(input: Pair<String, String> /*first - i, second - expression*/): Pair<String, String> /* first - i, second - result*/ {
        val list = ArrayList<String>()
        var number = "" // if we have a part of number, we`ll add it there and get a full number after all
        var i = input.first.toInt()
        val str = input.second
        while (i < str.length) {
            if (str[i] == ')') {
                i++
                break
            }
            if (str[i] != '(') {
                when (str[i]) {
                    '+', '-', '×', '÷', '^' -> {
                        list.add(str[i].toString())
                    }
                    else -> {
                        try {
                            if (str[i + 1] != '.')
                                Integer.parseInt(str[i + 1].toString())
                            // if next char is digit
                            number += str[i]
                        } catch (e: Exception) {
                            // if next char isn`t digit
                            number += str[i]
                            list.add(number)
                            number = ""
                        }
                    }
                }
                i++
            } else {
                val output = core(Pair((i + 1).toString(), str))
                i = output.first.toInt()
                list.add(output.second)
            }
        }

        // calculating from list
        var iterator = 0
        var a: Double
        var b: Double
        // ^ firstly
        while (true) {
            if (iterator == list.size) {
                iterator = 0
                break
            }
            if (list[iterator] == "^") {
                b = (list.removeAt(iterator + 1)).toDouble()
                list.removeAt(iterator)
                a = (list.removeAt(iterator - 1)).toDouble()
                list.add(iterator - 1, (a.pow(b)).toString())
                iterator = 0
            }
            iterator++;
        }
        // × and ÷
        while (true) {
            if (iterator == list.size) {
                iterator = 0
                break
            }
            if (list[iterator] == "×") {
                b = (list.removeAt(iterator + 1)).toDouble()
                list.removeAt(iterator)
                a = (list.removeAt(iterator - 1)).toDouble()
                list.add(iterator - 1, (a * b).toString())
                iterator = 0
            }
            if (list[iterator] == "÷") {
                b = (list.removeAt(iterator + 1)).toDouble()
                list.removeAt(iterator)
                a = (list.removeAt(iterator - 1)).toDouble()
                list.add(iterator - 1, (a / b).toString())
                iterator = 0
            }
            iterator++;
        }
        // + and -
        while (true) {
            if (iterator == list.size) {
                iterator = 0
                break
            }
            if (list[iterator] == "+") {
                b = (list.removeAt(iterator + 1)).toDouble()
                list.removeAt(iterator)
                a = (list.removeAt(iterator - 1)).toDouble()
                list.add(iterator - 1, (a + b).toString())
                iterator = 0
            }
            if (list[iterator] == "-") {
                b = (list.removeAt(iterator + 1)).toDouble()
                list.removeAt(iterator)
                a = (list.removeAt(iterator - 1)).toDouble()
                list.add(iterator - 1, (a - b).toString())
                iterator = 0
            }
            iterator++;
        }

        return Pair(i.toString(), list.removeAt(0))
    }

}